transcendence-app/src/game/game.service.ts

```diff
@@ -0,0 +1,343 @@
+  playerTwoLeftAt: number | null;
+};
+
+const FPS = 30;
+const DELTA_TIME = 1 / FPS;
+const MAX_PAUSED_TIME_MS = 30 * 1000; // 30 seconds
+
+@Injectable()
+export class GameService {
+  public server: Server | null = null;
+  private readonly logger = new Logger(GameService.name);
+  private readonly games = new Map<GameId, GameInfo>();
+  private readonly userToGame = new Map<UserId, GameId>();
+  private readonly playerClients = new Map<UserId, Set<ClientId>>();
+  private intervalId: NodeJS.Timer | undefined = undefined;
+
+  // constructor() {}
+
+  private finishGame(gameInfo: GameInfo, gameId: string) {
+    if (!this.server) {
+      return;
+    }
+    this.server.to(gameId).emit('gameFinished', gameInfo);
+    this.server
+      .to([gameInfo.playerOneId, gameInfo.playerTwoId])
+      .emit('gameStatusUpdate', {
+        status: GameStatus.FINISHED,
+      } as GameStatusUpdateDto);
+    this.server.socketsLeave(gameId);
+    this.playerClients.delete(gameInfo.playerOneId);
+    this.playerClients.delete(gameInfo.playerTwoId);
+    this.userToGame.delete(gameInfo.playerOneId);
+    this.userToGame.delete(gameInfo.playerTwoId);
+    this.games.delete(gameId);
+    if (this.games.size === 0) {
+      clearInterval(this.intervalId);
+      this.intervalId = undefined;
+    }
+  }
+
+  private runServerGameFrame() {
+    this.games.forEach((gameInfo: GameInfo, gameId: GameId) => {
+      if (this.server) {
+        if (
+          gameInfo.pausedAt &&
+          Date.now() >= gameInfo.pausedAt + MAX_PAUSED_TIME_MS
+        ) {
+          this.finishGame(gameInfo, gameId);
+        } else if (gameInfo.playState === 'playing') {
+          // Game is playing, update the game info
+          const gameState = runGameMultiplayerFrame(
+            DELTA_TIME,
+            gameInfo.gameState,
+          );
+          const newGameInfo: GameInfo = { ...gameInfo, gameState };
+          this.games.set(gameId, newGameInfo);
+          this.server.to(gameId).emit('updateGame', newGameInfo);
+        } else {
+          // Game is paused, send the current game info
+          this.server.to(gameId).emit('updateGame', gameInfo);
+        }
+      }
+    });
+  }
+
+  handleGameJoin(client: Socket, gameRoomId: string) {
+    if (!this.server) {
+      return;
+    }
+    const game = this.games.get(gameRoomId);
+    if (!game) {
+      this.server.to(client.id).emit('gameNotFound');
+      throw new WsException('Game not found');
+    }
+    client.join(gameRoomId);
+    this.server.to(client.id).emit('gameJoined', game);
+
+    const userId = client.request.user.id;
+    const isPlayerOne = userId === game.playerOneId;
+    const isPlayerTwo = userId === game.playerTwoId;
+
+    if (!(isPlayerOne || isPlayerTwo)) {
+      return;
+    }
+    const playerClients = this.playerClients.get(userId);
+    if (!playerClients) {
+      this.playerClients.set(userId, new Set<ClientId>([client.id]));
+    } else {
+      playerClients.add(client.id);
```

Comment:
should this be
```
      this.playerClients.get(userId)?.add(client.id);
```
Or is `playerClients` a reference to `this.playerClients.get(userId)` and any change on the first affects the seconds? I get confused with these things in javascript...

Reply:
Checked, this seems to work. Sorry for my ignorance.

Reply:
Sorry, this was a bit confusing. I renamed `this.playerClients` to `this.playerToClients`.

And yes, [`Map.prototype.get()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map/get) returns a reference, in this case, a reference to the player clients set or `undefined`.